src/core/types/fee_payment_detail.rs

```diff
@@ -138,16 +157,82 @@ fn generate_verifier_fee_name(verifier: &VerifierDetailV2) -> String {
         .unwrap_or_else(|| "Verifier Fee".to_string())
 }
 
+fn calc_onboarding_cost_source<S: Into<String>>(
+    verifier: &VerifierDetailV2,
+    is_retry: bool,
+    asset_type: S,
+    existing_scope_attributes: &[AssetScopeAttribute],
+) -> AssetResult<OnboardingCost> {
+    let asset_type = asset_type.into();
+    // Fetch all scope attributes on the asset that used this verifier and were not for this target
+    // asset type.  If this is not empty, that means that this request is a subsequent classification
+    // for the same verifier and can use subsequent costs.
+    let other_classifications = existing_scope_attributes
+        .iter()
+        .filter(|attr| {
+            attr.verifier_address.as_str() == verifier.address && attr.asset_type != asset_type
+        })
+        .collect::<Vec<&AssetScopeAttribute>>();
+    // Always favor retry cost.  Regardless of the scenario, retries should override the specified
+    // root costs and/or subsequent classification costs
+    if is_retry {
+        verifier.get_retry_cost().to_ok()
+    } else if !other_classifications.is_empty() {
+        // If at least one other classification that used this verifier is present in the existing
+        // scope attributes, then this qualifies as a subsequent classification.
+        if let Some(ref allowed_types) = verifier
+            .subsequent_classification_detail
+            .clone()
+            .and_then(|d| d.allowed_asset_types)
+        {
+            let classified_asset_types = other_classifications
+                .iter()
+                .cloned()
+                .map(|attr| attr.asset_type.to_owned())
+                .collect::<Vec<String>>();
+            // If the allowed asset types node is set, and at least one type that this asset is
+            // classified as is not present in that vector, then this is an unsupported
+            // subsequent classification with this verifier and the request needs to be rejected.
+            if !classified_asset_types
+                .iter()
+                .all(|asset_type| allowed_types.contains(&asset_type))
+            {
+                return ContractError::UnsupportedSubsequentAssetType {
```

Comment:
I guess I was envisioning that the allowed asset type list would just serve to determine whether the subsequent classification fee applies or just the default root fee applies, rather than serve to outright reject the attempt... thoughts?

Reply:
Fixed 🖖